generate/post_process_json.go

```diff
@@ -0,0 +1,113 @@
+package generate
+
+import (
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"reflect"
+	"strings"
+)
+
+// jsonNameFromProtobufTag parse json name from protobuf tag, returns empty string if no json name is defined.
+// example: protobuf:"bytes,1,opt,name=FirstName,json=first_name,proto3"
+// result: first_name
+func jsonNameFromProtobufTag(tag string) string {
+	const jsonPartName = "json="
+	if len(tag) == 0 {
+		return ""
+	}
+	parts := strings.Split(tag, ",")
+	for _, part := range parts {
+		if strings.HasPrefix(part, jsonPartName) {
+			return strings.TrimPrefix(part, jsonPartName)
+		}
+	}
+	return ""
+}
+
+// jsonNameFromJSONTag returns the json name from json tag, returns empty string if not defined.
+func jsonNameFromJSONTag(tag string) string {
+	if len(tag) == 0 {
+		return ""
+	}
+	parts := strings.Split(tag, ",")
+	if len(parts) == 0 {
+		return ""
+	}
+	return parts[0]
+}
+
+type textArea struct {
+	pos  token.Pos
+	end  token.Pos
+	text string
+}
+
+func jsonTagPostProcessor(input []byte) ([]byte, error) {
+	const (
+		jsonTagName     = "json"
+		protobufTagName = "protobuf"
+	)
+	var textAreas []textArea
+
+	fset := token.NewFileSet()
+	f, err := parser.ParseFile(fset, "", input, parser.ParseComments)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, decl := range f.Decls {
+		// check if is generic declaration
+		genDecl, ok := decl.(*ast.GenDecl)
+		if !ok {
+			continue
+		}
+
+		var typeSpec *ast.TypeSpec
+		for _, spec := range genDecl.Specs {
+			if ts, tsOK := spec.(*ast.TypeSpec); tsOK {
+				typeSpec = ts
+				break
+			}
+		}
+
+		// skip if can't get type spec
+		if typeSpec == nil {
+			continue
+		}
+
+		// not a struct, skip
+		structDecl, ok := typeSpec.Type.(*ast.StructType)
+		if !ok {
+			continue
+		}
+
+		for _, field := range structDecl.Fields.List {
```

Comment:
You can replace all of this with `ast.Inspect`, and then filtering by the `*ast.StructType` type.

Reply:
replaced the loop with ast.Inspect, that looks so much better. Thank for suggestion.